go/types/struct.go

```diff
@@ -169,18 +169,33 @@ func (s1 Struct) Diff(s2 Struct, changes chan<- ValueChanged, closeChan <-chan s
 }
 
 var escapeChar = "Q"
-var headPattern = regexp.MustCompile("[a-zA-PR-Z]")
-var tailPattern = regexp.MustCompile("[a-zA-PR-Z0-9_]")
-var completePattern = regexp.MustCompile("^" + headPattern.String() + tailPattern.String() + "*$")
+var headFieldNamePattern = regexp.MustCompile("[a-zA-Z]")
+var tailFieldNamePattern = regexp.MustCompile("[a-zA-Z0-9_]")
+var escapeRegex = regexp.MustCompile(escapeChar)
+
+var fieldNameComponentRe = regexp.MustCompile("^" + headFieldNamePattern.String() + tailFieldNamePattern.String() + "*")
+var fieldNameRe = regexp.MustCompile(fieldNameComponentRe.String() + "$")
+
+type encodingFunc func(string, *regexp.Regexp) string
```

Comment:
I see why you are doing this, but I find the signature awkward. As a caller of `EscapeFields`, how do I know what to do with the regex that is passed to `encodingFunc`.

Could you factor this like so instead?

`func QEncodeFieldName(name string) string`

and:

`func CamelCaseFieldName(name string) string`

all the rest of the guts should be package-private implementation details, not exposed to caller.